File: easy-random-bean-validation/src/main/java/org/jeasy/random/validation/AbstractTemporalBaseAnnotationHandler.java
package org.jeasy.random.validation;

import org.jeasy.random.EasyRandom;
import org.jeasy.random.api.Randomizer;
import org.jeasy.random.randomizers.range.*;

import java.lang.reflect.Field;
import java.time.*;
import java.time.chrono.HijrahDate;
import java.time.chrono.JapaneseDate;
import java.time.chrono.MinguoDate;
import java.time.chrono.ThaiBuddhistDate;
import java.util.Calendar;
import java.util.Date;

public abstract class AbstractTemporalBaseAnnotationHandler
    implements BeanValidationAnnotationHandler {
  protected final Date minDate;
  protected final Date maxDate;
  protected final Instant minInstant;
  protected final Instant maxInstant;
  protected final LocalDate minLocalDate;
  protected final LocalDate maxLocalDate;
  protected final LocalDateTime minLocalDateTime;
  protected final LocalDateTime maxLocalDateTime;
  protected final LocalTime minLocalTime;
  protected final LocalTime maxLocalTime;
  protected final MonthDay minMonthDay;
  protected final MonthDay maxMonthDay;
  protected final OffsetDateTime minOffsetDateTime;
  protected final OffsetDateTime maxOffsetDateTime;
  protected final OffsetTime minOffsetTime;
  protected final OffsetTime maxOffsetTime;
  protected final Year minYear;
  protected final Year maxYear;
  protected final YearMonth minYearMonth;
  protected final YearMonth maxYearMonth;
  protected final ZonedDateTime minZonedDateTime;
  protected final ZonedDateTime maxZonedDateTime;

  protected AbstractTemporalBaseAnnotationHandler(
      Date minDate,
      Date maxDate,
      Instant minInstant,
      Instant maxInstant,
      LocalDate minLocalDate,
      LocalDate maxLocalDate,
      LocalDateTime minLocalDateTime,
      LocalDateTime maxLocalDateTime,
      LocalTime minLocalTime,
      LocalTime maxLocalTime,
      MonthDay minMonthDay,
      MonthDay maxMonthDay,
      OffsetDateTime minOffsetDateTime,
      OffsetDateTime maxOffsetDateTime,
      OffsetTime minOffsetTime,
      OffsetTime maxOffsetTime,
      Year minYear,
      Year maxYear,
      YearMonth minYearMonth,
      YearMonth maxYearMonth,
      ZonedDateTime minZonedDateTime,
      ZonedDateTime maxZonedDateTime) {
    this.minDate = minDate;
    this.maxDate = maxDate;
    this.minInstant = minInstant;
    this.maxInstant = maxInstant;
    this.minLocalDate = minLocalDate;
    this.maxLocalDate = maxLocalDate;
    this.minLocalDateTime = minLocalDateTime;
    this.maxLocalDateTime = maxLocalDateTime;
    this.minLocalTime = minLocalTime;
    this.maxLocalTime = maxLocalTime;
    this.minMonthDay = minMonthDay;
    this.maxMonthDay = maxMonthDay;
    this.minOffsetDateTime = minOffsetDateTime;
    this.maxOffsetDateTime = maxOffsetDateTime;
    this.minOffsetTime = minOffsetTime;
    this.maxOffsetTime = maxOffsetTime;
    this.minYear = minYear;
    this.maxYear = maxYear;
    this.minYearMonth = minYearMonth;
    this.maxYearMonth = maxYearMonth;
    this.minZonedDateTime = minZonedDateTime;
    this.maxZonedDateTime = maxZonedDateTime;
  }

  @Override
  public Randomizer<?> getRandomizer(Field field) {
    if (field.getType().equals(Date.class)) {
      return () -> new DateRangeRandomizer(minDate, maxDate).getRandomValue();
    } else if (field.getType().equals(Calendar.class)) {
      return () -> {
        Calendar calendar = Calendar.getInstance();
        Date date = new DateRangeRandomizer(minDate, maxDate).getRandomValue();
        calendar.setTime(date);
        return calendar;
      };
    } else if (field.getType().equals(Instant.class)) {
      return () -> new InstantRangeRandomizer(minInstant, maxInstant).getRandomValue();
    } else if (field.getType().equals(LocalDate.class)) {
      return () -> new LocalDateRangeRandomizer(minLocalDate, maxLocalDate).getRandomValue();
    } else if (field.getType().equals(LocalDateTime.class)) {
      return () ->
          new LocalDateTimeRangeRandomizer(minLocalDateTime, maxLocalDateTime).getRandomValue();
    } else if (field.getType().equals(LocalTime.class)) {
      return () -> new LocalTimeRangeRandomizer(minLocalTime, maxLocalTime).getRandomValue();
    } else if (field.getType().equals(MonthDay.class)) {
      return () -> new MonthDayRangeRandomizer(minMonthDay, maxMonthDay).getRandomValue();
    } else if (field.getType().equals(OffsetDateTime.class)) {
      return () ->
          new OffsetDateTimeRangeRandomizer(minOffsetDateTime, maxOffsetDateTime).getRandomValue();
    } else if (field.getType().equals(OffsetTime.class)) {
      return () -> new OffsetTimeRangeRandomizer(minOffsetTime, maxOffsetTime).getRandomValue();
    } else if (field.getType().equals(Year.class)) {
      return () -> new YearRangeRandomizer(minYear, maxYear).getRandomValue();
    } else if (field.getType().equals(YearMonth.class)) {
      return () -> new YearMonthRangeRandomizer(minYearMonth, maxYearMonth).getRandomValue();
    } else if (field.getType().equals(ZonedDateTime.class)) {
      return () ->
          new ZonedDateTimeRangeRandomizer(minZonedDateTime, maxZonedDateTime).getRandomValue();
    } else if (field.getType().equals(HijrahDate.class)) {
      return () ->
          HijrahDate.from(
              new LocalDateRangeRandomizer(minLocalDate, maxLocalDate).getRandomValue());
    } else if (field.getType().equals(JapaneseDate.class)) {
      return () ->
          JapaneseDate.from(
              new LocalDateRangeRandomizer(minLocalDate, maxLocalDate).getRandomValue());
    } else if (field.getType().equals(MinguoDate.class)) {
      return () ->
          MinguoDate.from(
              new LocalDateRangeRandomizer(minLocalDate, maxLocalDate).getRandomValue());
    } else if (field.getType().equals(ThaiBuddhistDate.class)) {
      return () ->
          ThaiBuddhistDate.from(
              new LocalDateRangeRandomizer(minLocalDate, maxLocalDate).getRandomValue());
    } else {
      return () -> new EasyRandom().nextObject(field.getType());
    }
  }
}

File: bean-validation/src/main/kotlin/io/github/krandom/validation/AbstractTemporalBaseAnnotationHandler.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.github.krandom.api.Randomizer
import io.github.krandom.randomizers.range.DateRangeRandomizer
import io.github.krandom.randomizers.range.InstantRangeRandomizer
import io.github.krandom.randomizers.range.LocalDateRangeRandomizer
import io.github.krandom.randomizers.range.LocalDateTimeRangeRandomizer
import io.github.krandom.randomizers.range.LocalTimeRangeRandomizer
import io.github.krandom.randomizers.range.MonthDayRangeRandomizer
import io.github.krandom.randomizers.range.OffsetDateTimeRangeRandomizer
import io.github.krandom.randomizers.range.OffsetTimeRangeRandomizer
import io.github.krandom.randomizers.range.YearMonthRangeRandomizer
import io.github.krandom.randomizers.range.YearRangeRandomizer
import io.github.krandom.randomizers.range.ZonedDateTimeRangeRandomizer
import java.lang.reflect.Field
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.MonthDay
import java.time.OffsetDateTime
import java.time.OffsetTime
import java.time.Year
import java.time.YearMonth
import java.time.ZonedDateTime
import java.time.chrono.HijrahDate
import java.time.chrono.JapaneseDate
import java.time.chrono.MinguoDate
import java.time.chrono.ThaiBuddhistDate
import java.util.Calendar
import java.util.Date
import kotlin.getValue

/** @author Alex Schiff */
open class AbstractTemporalBaseAnnotationHandler
protected constructor(protected val zonedDateTimeRange: ClosedRange<ZonedDateTime>) :
  BeanValidationAnnotationHandler {
  private val randomizerByType: Map<Class<*>, Randomizer<*>> by lazy {
    val date =
      DateRangeRandomizer(
          Date.from(zonedDateTimeRange.start.toInstant()),
          Date.from(zonedDateTimeRange.endInclusive.toInstant()),
        )
        .getRandomValue()
    mapOf(
      Date::class.java to Randomizer { date },
      Calendar::class.java to
        Randomizer {
          val calendar = Calendar.getInstance()
          calendar.time = date
          calendar
        },
      Instant::class.java to
        Randomizer {
          InstantRangeRandomizer(
              zonedDateTimeRange.start.toInstant(),
              zonedDateTimeRange.endInclusive.toInstant(),
            )
            .getRandomValue()
        },
      LocalDate::class.java to
        Randomizer {
          LocalDateRangeRandomizer(
              zonedDateTimeRange.start.toLocalDate(),
              zonedDateTimeRange.endInclusive.toLocalDate(),
            )
            .getRandomValue()
        },
      LocalDateTime::class.java to
        Randomizer {
          LocalDateTimeRangeRandomizer(
              zonedDateTimeRange.start.toLocalDateTime(),
              zonedDateTimeRange.endInclusive.toLocalDateTime(),
            )
            .getRandomValue()
        },
      LocalTime::class.java to
        Randomizer {
          LocalTimeRangeRandomizer(
              zonedDateTimeRange.start.toLocalTime(),
              zonedDateTimeRange.endInclusive.toLocalTime(),
            )
            .getRandomValue()
        },
      MonthDay::class.java to
        Randomizer {
          MonthDayRangeRandomizer(
              MonthDay.from(zonedDateTimeRange.start),
              MonthDay.from(zonedDateTimeRange.endInclusive),
            )
            .getRandomValue()
        },
      OffsetDateTime::class.java to
        Randomizer {
          OffsetDateTimeRangeRandomizer(
              zonedDateTimeRange.start.toOffsetDateTime(),
              zonedDateTimeRange.endInclusive.toOffsetDateTime(),
            )
            .getRandomValue()
        },
      OffsetTime::class.java to
        Randomizer {
          OffsetTimeRangeRandomizer(
              zonedDateTimeRange.start.toOffsetDateTime().toOffsetTime(),
              zonedDateTimeRange.endInclusive.toOffsetDateTime().toOffsetTime(),
            )
            .getRandomValue()
        },
      Year::class.java to
        Randomizer {
          YearRangeRandomizer(
              Year.from(zonedDateTimeRange.start),
              Year.from(zonedDateTimeRange.endInclusive),
            )
            .getRandomValue()
        },
      YearMonth::class.java to
        Randomizer {
          YearMonthRangeRandomizer(
              YearMonth.from(zonedDateTimeRange.start),
              YearMonth.from(zonedDateTimeRange.endInclusive),
            )
            .getRandomValue()
        },
      ZonedDateTime::class.java to
        Randomizer {
          ZonedDateTimeRangeRandomizer(zonedDateTimeRange.start, zonedDateTimeRange.endInclusive)
            .getRandomValue()
        },
      HijrahDate::class.java to
        Randomizer {
          HijrahDate.from(
            LocalDateRangeRandomizer(
                zonedDateTimeRange.start.toLocalDate(),
                zonedDateTimeRange.endInclusive.toLocalDate(),
              )
              .getRandomValue()
          )
        },
      JapaneseDate::class.java to
        Randomizer {
          JapaneseDate.from(
            LocalDateRangeRandomizer(
                zonedDateTimeRange.start.toLocalDate(),
                zonedDateTimeRange.endInclusive.toLocalDate(),
              )
              .getRandomValue()
          )
        },
      MinguoDate::class.java to
        Randomizer {
          MinguoDate.from(
            LocalDateRangeRandomizer(
                zonedDateTimeRange.start.toLocalDate(),
                zonedDateTimeRange.endInclusive.toLocalDate(),
              )
              .getRandomValue()
          )
        },
      ThaiBuddhistDate::class.java to
        Randomizer {
          ThaiBuddhistDate.from(
            LocalDateRangeRandomizer(
                zonedDateTimeRange.start.toLocalDate(),
                zonedDateTimeRange.endInclusive.toLocalDate(),
              )
              .getRandomValue()
          )
        },
    )
  }

  override fun getRandomizer(field: Field): Randomizer<*>? =
    randomizerByType[field.type] ?: Randomizer { KRandom().nextObject(field.type) }
}
